### ISSUE TITLE
Vulkan 1.0 doesn't consider deviceCoherentMemory to be an extension

### ISSUE THREAD

**Issue by kvark** [KhronosGroup/Vulkan-Docs#1298]
We run on Vulkan 1.0 instance, we don't enable "VK_AMD_device_coherent_memory" extension, and we get this validation error (see https://github.com/gfx-rs/wgpu/issues/718):
> VALIDATION [VUID-vkAllocateMemory-deviceCoherentMemory-02790 (-2010063723)] : Validation Error: [ VUID-vkAllocateMemory-deviceCoherentMemory-02790 ] Object 0: handle = 0x55b416ae07b0, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x8830dc95 | vkAllocateMemory: attempting to allocate memory type 4, which includes the VK_MEMORY_PROPERTY_DEVICE_COHERENT_BIT_AMD memory property, but the deviceCoherentMemory feature is not enabled. The Vulkan spec states: If the deviceCoherentMemory feature is not enabled, pAllocateInfo->memoryTypeIndex must not identify a memory type supporting VK_MEMORY_PROPERTY_DEVICE_COHERENT_BIT_AMD (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkAllocateMemory-deviceCoherentMemory-02790)
    object info: (type: DEVICE, hndl: 94231962978224)

The VU statement in 1.0 [states](https://www.khronos.org/registry/vulkan/specs/1.0-extensions/html/vkspec.html#VUID-vkAllocateMemory-deviceCoherentMemory-02790):
> If the `deviceCoherentMemory` feature is not enabled, `pAllocateInfo->memoryTypeIndex` must not identify a memory type supporting VK_MEMORY_PROPERTY_DEVICE_COHERENT_BIT_AMD

And it looks like this is precisely what the validation layer implements. Both appear to be a mistake to me, since the spec can only add this requirement if "VK_AMD_device_coherent_memory" extension is enabled.

Is this intentional, or just an omission, that the extension presence is not mentioned `deviceCoherentMemory` VU statement?

**Comment by krOoze** [comment]
#533 strikes again!

**Comment by kvark** [comment]
Thanks @krOoze , this is indeed another manifestation of this old unresolved issue!
I wouldn't care that much, but the validation layers are complaining about this, and it annoys me and our users quite a bit :(

**Comment by krOoze** [comment]
The error seems directionally correct though. I am pretty sure it is bad to allocate from these memory types unless you enable the extension.

**Comment by kvark** [comment]
@krOoze I don't see how this would work. Whenever a driver implements an extension or a higher Vulkan version, it has to keep all the existing applications running. If there is an extension for "SUPER_FAST_MEMORY", and the driver happens to support it universally, putting the bit on all memory types, an application that filters out the types would end up broken.

Instead, I believe it's more reasonable to expect the driver to only expose the bits that correspond to enabled extensions / vulkan version. I.e. in our case the driver should realize that it can't expose any memory types with this bit.

**Comment by krOoze** [comment]
The driver cannot do it neither, as extensions are enabled at `vkCreateDevice` time, but memory types are queried from `VkPhysicalDevice`.

**Comment by kvark** [comment]
I see... Maybe that means the extension needs to really be an instance extension instead of the device one?

**Comment by Tobski** [comment]
> I see... Maybe that means the extension needs to really be an instance extension instead of the device one?

That ship has kind of sailed, and we did have this discussion before we shipped it. Making it an instance extension would've come with other problems as well...

So we instead avoided breaking shipping applications by adhering to the "slower memory shows up last" principle (consistent with recommended allocation strategies), and adding this VU so that applications would have validation errors if they didn't follow that.

This line:

> Any Vk*Flags member or parameter returned from a query command or otherwise output from Vulkan to the application may contain bit flags undefined in its corresponding Vk*FlagBits type. An application cannot rely on the state of these unspecified bits.

Is kind of suggesting that applications only deal with things they know about - format features work on the same principle for instance. If you see something you don't understand, you don't use it, in whatever capacity that takes shape.

In this case it's a bit weird because there's nothing to enable based on the returned bit - it means just not using memory types with unknown memory properties.

> @krOoze I don't see how this would work. Whenever a driver implements an extension or a higher Vulkan version, it has to keep all the existing applications running. If there is an extension for "SUPER_FAST_MEMORY", and the driver happens to support it universally, putting the bit on all memory types, an application that filters out the types would end up broken.

I mean that's kind of what we did? That was actually a consideration when we specced this extension - we considered saying "existing memory types are CACHED" and then adding new types which were not cached. One of the reasons we specifically didn't do that was because it would've broken shipping apps that didn't use an unknown memory type, so they would've ended up on the slow path. For similar reasons, I expect that any such hypothetical extension would thus be forced into advertising a "SUPER_SLOW_MEMORY_BIT" instead (i.e. negating the meaning), and putting these on additional memory types.

**Comment by kvark** [comment]
Thank you for clarification @Tobski !
I understand that the situation here was complex, and that the group considered all the outcomes. I think there is still a problem with specification today.

> So we instead avoided breaking shipping applications by adhering to the "slower memory shows up last" principle (consistent with recommended allocation strategies), and adding this VU so that applications would have validation errors if they didn't follow that.

So suddenly a recommendation became a VU requirement. That's a breaking change.

>> Any VkFlags member or parameter returned from a query command or otherwise output from Vulkan to the application may contain bit flags undefined in its corresponding VkFlagBits type. An application cannot rely on the state of these unspecified bits.
> Is kind of suggesting that applications only deal with things they know about - format features work on the same principle for instance. If you see something you don't understand, you don't use it, in whatever capacity that takes shape.

I don't want the spec to be suggestive. I want it to be precise. Am I allowed to use the memory types with bits that I don't know about? Current spec doesn't say NO. It says only that I can't rely on the state of these bits. Since I don't know about these bits, I don't rely on their state. So the spec here is a bit of a nonsense.

Can we write the spec saying that the app *is not allowed* to use memory types with unknown bits?

**Comment by Ralith** [comment]
I agree this looks like a spec oversight. Yes, you can infer that using a memory type with an unknown flag could run afoul of unknown requirements, but the presence of that hazard is far from obvious. So non-obvious, in fact, that the `findProperties` example search procedure seems to have a related bug: if `requiredProperties` is only satisfied by a memory type which additionally has an unknown property, that memory type will be selected.

**Comment by krOoze** [comment]
Feels little bit related to the versioning issues. The spec should be clearer and more careful to specify what app should (and shouldn't) do to achieve forwards-compatibility with new versions and new extensions.

**Comment by Tobski** [comment]
> I don't want the spec to be suggestive. I want it to be precise. Am I allowed to use the memory types with bits that I don't know about? Current spec doesn't say NO. It says only that I can't rely on the state of these bits. Since I don't know about these bits, I don't rely on their state. So the spec here is a bit of a nonsense.
> 
> Can we write the spec saying that the app is not allowed to use memory types with unknown bits?

Ok so I dived into this again, and AFAICT the spec and validation are currently doing "the right thing" here. The language quoted about flag bits now tells you that a flag bit is only valid if it's a valid enum for the type, and an enum that is not part of an enabled version or extension is not valid - so the circle here seems to be closed and complete? The validation layers are also (to my knowledge) definitively catching this.

Device coherent memory was a useful canary because there are other parts of the spec which do stuff like this, just most of them don't get used quite as heavily as memory allocation...

So I'm going to close this, but there was quite a lot going on here so feel free to reopen if I've missed something.
